### PR TITLE
Support layered `type_encoders`

### PR DESCRIPTION
**Pull request by provinzkraut**
# PR Checklist

- [x] Have you followed the guidelines in `CONTRIBUTING.md`?
- [x] Have you got 100% test coverage on new code?
- [x] Have you updated the prose documentation?
- [x] Have you updated the reference documentation?

### CODE REVIEW

File: starlite/handlers/http.py [hunk 1]
```diff
@@ -563,6 +570,20 @@ def resolve_after_response(self) -> Optional["AfterResponseHookHandler"]:
 
         return cast("Optional[AfterResponseHookHandler]", self._resolved_after_response)
 
+    def resolve_type_encoders(self) -> Optional[TypeEncodersMap]:
+        """Resolve `type_encoders` by merging existing `type_encoders` from all layers.
+
+        Returns:
+            A `TypeEncodersMap` to use for this response or `None`
+        """
+        type_encoders: Optional[TypeEncodersMap] = None
```
Comment (Goldziher) [inline starlite/handlers/http.py:579]: you can default this to be {} and do an OR in the return statement.

**Review by Goldziher — APPROVED** [review]

File: starlite/types/composite_types.py [hunk 2]
```diff
@@ -59,3 +59,4 @@
     Callable[..., ASGIApp], DefineMiddleware, Iterator[Tuple[ASGIApp, Dict[str, Any]]], Type[MiddlewareProtocol]
 ]
 MaybePartial = Union[T, partial]
+TypeEncodersMap = Dict[Any, Callable[[Any], Any]]
```
Comment (Goldziher) [inline starlite/types/composite_types.py:62]: Should the callables return any or bytes?

**Review by Goldziher — APPROVED** [review]

File: starlite/types/composite_types.py (hunk 2 above)
Comment (provinzkraut) [reply starlite/types/composite_types.py:62]: `Any` is correct here. They only transform values into types that `msgspec` can understand, they don't serialize anything themselves.

File: starlite/response/base.py [hunk 3]
```diff
@@ -119,6 +125,40 @@ def __init__(
...
+        "1.48.0",
+        pending=True,
+        info="Set `type_encoders` on a higher layer "
+        "instead or pass `type_encoders` to the `Response` instance directly",
+    )
+    def serializer(cls, value: Any) -> Any:
+        """Transform non-natively supported types into supported types.
+
+        Should raise `TypeError` if a type cannot be transformed into a supported type
+        """
```
Comment (peterschutt) [inline starlite/response/base.py:139]: Hey @provinzkraut - as this is meant to be a deprecation, shouldn't this still return a serialized value?

Comment (provinzkraut) [reply starlite/response/base.py:139]: Nope. This method isn’t actually being used anymore by the serializer, unless you override it in a subclass. Look at the `get_serializer` method.

This is done to provide an easier upgrade path for us, avoid passing around the default serializer, and a slight performance benefit.

Comment (peterschutt) [reply starlite/response/base.py:139]: But `Response.serializer()` is a documented method, and this changes its behavior.

E.g., this worked fine in the previous version:

```python
def default_serializer(value: Any) -> Any:
    """Serialize `value`.
    Args:
        value: To be serialized.
    Returns:
        Serialized representation of `value`.
    """
    if isinstance(value, pgproto.UUID):
        return str(value)
    return Response[Any].serializer(value)
```
Not a big issue, as now that I've pinpointed what is wrong I'm updating to the layered type encoders (which is a much better pattern), just wanted to check that this isn't an over site.

Smth similar going on in `starlite-pg-redis-docker` app, which I'll update now in due course.

Comment (provinzkraut) [reply starlite/response/base.py:139]: Ah, you’re right. I actually hadn’t thought of this case. Guess we can just revert to calling `default_serializer` in `serializer` to re-enable this. 
I’ll fix this tomorrow!

Comment (provinzkraut) [reply starlite/response/base.py:139]: @peterschutt Should be fixed by #1018

Comment (peterschutt) [reply starlite/response/base.py:139]: Thanks!

Comment (provinzkraut) [reply starlite/response/base.py:139]: Just needs a review so I can merge it :wink:

**Pull request merged**